clients/cobol-lsp-vscode-extension/src/dialect/DialectService.ts

```diff
@@ -0,0 +1,343 @@
+import { LanguageClientService } from "../services/LanguageClientService";
+import { Location } from "vscode";
+import { CopybookDownloadService } from "../services/copybook/CopybookDownloadService";
+import { readFileContent } from "../services/copybook/CopybookMessageHandler";
+
+type CopybookPayload = {
+  copybookName: string;
+  statementLocation: unknown;
+  nameLocation: unknown;
+  uri: string;
+  text: string;
+  replacements: unknown[];
+  copybooks: CopybookPayload[];
+};
+
+type DocumentReplacement = {
+  range: vscode.Range;
+  text: string;
+};
+
+type CopybookInfo = {
+  copybookName: string;
+  statementLocation: Location;
+  nameLocation: Location;
+  uri: string;
+  text: string;
+  replacements: DocumentReplacement[];
+  copybooks: Map<string, CopybookInfo>;
+};
+
+class Context implements IDocumentProcessingContext {
+  constructor(
+    private dialectService: DialectService,
+    private dialectName: string,
+    private programUri: string,
+    private documentUri: string,
+    private parentCopybook?: CopybookInfo,
+  ) {}
+
+  async resolveCopybook(
+    copybookName: string,
+    statementRange: vscode.Range,
+    nameRange: vscode.Range,
+  ): Promise<
+    | {
+        context: IDocumentProcessingContext;
+        uri: string;
+        text: string;
+      }
+    | undefined
+  > {
+    const statementLocation = new vscode.Location(
+      vscode.Uri.parse(this.documentUri),
+      statementRange,
+    );
+    const nameLocation = new vscode.Location(
+      vscode.Uri.parse(this.documentUri),
+      nameRange,
+    );
+
+    const copybookInfo = await this.dialectService.resolveCopybook(
+      this.dialectName,
+      this.programUri,
+      copybookName,
+      statementLocation,
+      nameLocation,
+      this.parentCopybook,
+    );
+    if (copybookInfo) {
+      const copybookContext = new Context(
+        this.dialectService,
+        this.dialectName,
+        this.programUri,
+        copybookInfo.uri,
+        copybookInfo,
+      );
+      return {
+        context: copybookContext,
+        uri: copybookInfo.uri,
+        text: copybookInfo.text,
+      };
+    }
+    return undefined;
+  }
+  replace(range: vscode.Range, text: string): void {
+    const location = new Location(vscode.Uri.parse(this.documentUri), range);
+    this.dialectService.replace(
+      this.dialectName,
+      this.programUri,
+      location,
+      text,
+      this.parentCopybook,
+    );
+  }
+}
+
+export class DialectService {
+  private handlers: Map<string, V2StartProcessingHandler> = new Map();
+  private replacements: Map<string, DocumentReplacement[]> = new Map();
+  private copybooks: Map<string, Map<string, CopybookInfo>> = new Map();
+
+  private diagnosticService: DialectDiagnosticService =
+    new DialectDiagnosticService();
+
+  public constructor(
+    languageClientService: LanguageClientService,
+    private copybookService: CopybookDownloadService,
+    private outputChannel: vscode.OutputChannel,
```

Comment:
No need to mock it in tests.
```suggestion
    private outputChannel?: vscode.OutputChannel,
```